packages/frontpage/app/(app)/post/[postAuthor]/[postRkey]/_lib/editor-state-to-comment-content.ts

```diff
@@ -0,0 +1,59 @@
+import {
+  $getRoot,
+  $isElementNode,
+  $isTextNode,
+  EditorState,
+  LexicalNode,
+} from "lexical";
+
+export function editorStateToCommentContent(editorState: EditorState) {
+  return editorState.read(() => {
+    const root = $getRoot();
+    const content = $nodeToCommentContent(root);
+    return content;
+  });
+}
+
+const FORMATS = ["bold", "italic", "strikethrough"] as const;
+
+type CommentFacet =
+  | {
+      $type: "fyi.frontpage.richtext.facet#format";
+      format: (typeof FORMATS)[number];
+    }
+  | {
+      $type: "fyi.frontpage.richtext.facet#link";
+      uri: string;
+    };
+
+type CommentContent = string | { content: string; facets: CommentFacet[] };
+
+function $nodeToCommentContent(node: LexicalNode): CommentContent[] {
```

Comment:
Some notes on the `ComentContent[]` type for later as this will be included in our lexicon:

There are two representations for some text without any facets:

- `"foo"`
- `{content: "foo", facets: []}`

I'm not sure there are good ways around this without making the type much harder to encode and decode.


We force an array type even though `CommentContent | CommentContent[]` would be equally valid and not much more difficult to implement. This is to make it easier to decode later.

This reduces the amount of possible empty value representations by one, but there are still a few. These are all equivalent and mean "empty value"

- empty array: `[]`
- empty array with string: `[""]`
- empty array with obj: `[{content: "", facets: []}]`
- empty array with many strings/objs: `["", {content: "", facets: []}]`

This actually leads to there being an infinite amount of possible empty values... This is really not ideal because it can't easily be caught by putting char length constraints on the text, as no matter how many nodes you have the length of the concatenated string will always be 0. I think this needs to be fixed somehow.

Reply:
Having multiple ways of doing something in the data model is bad because in general the more ways there are to do something, the more complex it is to enforce constraints, which leads to a higher likelihood of a loophole being possible.

Reply:
Worth noting that we're constrained by the atproto lexicon spec here, some solutions (such as non-empty array or string types) are not possible right now.

Reply:
I originally thought that a flat array would suitable here, but actually nested nodes need to be supported.

Consider a link that has some characters bolded, but not all. In markdown:

```md
[some **link**](example.com)
```

We don't want to represent this as a flat array because the best we can do with that is:

```md
[some ](example.com)[**link**](example.com)
```

This is not what the author intended. Adding heuristics to merge sibling links is not possible either because it should be valid to have two links that are the same next to eachother in the text.

The question now becomes: should rich text nodes be able to be nested arbitrarily or are there some rules we should put in place here to prevent arbitrary recursion depths?

Reply:
The nested structure also introduces another way to represent empty text:

```
[
  {
    content: [
      {
        content: [...],
        facets: []
    ],
    facets: []
  }
]
```